include/mc_state_observation/measurements/ContactsManager.hpp

```diff
@@ -0,0 +1,268 @@
+{
+  const auto & measRobot = ctl.robot(robotName);
+
+  contactsFound_.clear();
+
+  auto insert_contact = [this, &measRobot, onAddedContact](const std::string & surfaceName)
+  {
+    const auto & fs = measRobot.surfaceHasForceSensor(surfaceName) ? measRobot.surfaceForceSensor(surfaceName)
+                                                                   : measRobot.indirectSurfaceForceSensor(surfaceName);
+
+    ContactWithSensor & contactWS = addContactToManager(fs.name(), surfaceName, onAddedContact);
+    contactWS.forceNorm(fs.wrenchWithoutGravity(measRobot).force().norm());
+    if(contactWS.forceNorm() > contactDetectionThreshold_)
+    {
+      // the contact is added to the map of contacts using the name of the associated sensor
+      contactsFound_.insert(contactWS.id());
+    }
+  };
+
+  for(const auto & contact : ctl.solver().contacts())
+  {
+
+    const auto & r1 = ctl.robots().robot(contact.r1Index());
+    const auto & r2 = ctl.robots().robot(contact.r2Index());
+    if(r1.name() == measRobot.name())
+    {
+
+      if(r2.mb().nrDof() == 0) { insert_contact(contact.r1Surface()->name()); }
+    }
+    else if(r2.name() == measRobot.name())
+    {
+      if(r1.mb().nrDof() == 0) { insert_contact(contact.r1Surface()->name()); }
+    }
+  }
+}
+
+template<typename ContactT>
+void ContactsManager<ContactT>::findContactsFromSurfaces(const mc_control::MCController & ctl,
+                                                         const std::string & robotName)
+{
+  const auto & measRobot = ctl.robot(robotName);
+
+  contactsFound_.clear();
+
+  for(auto & contact : contacts())
+  {
+    const std::string & fsName = contact.second.forceSensor();
+    const mc_rbdyn::ForceSensor & forceSensor = measRobot.forceSensor(fsName);
+    contact.second.forceNorm(forceSensor.wrenchWithoutGravity(measRobot).force().norm());
+    if(contact.second.forceNorm() > contactDetectionThreshold_)
+    {
+      //  the contact is added to the map of contacts using the name of the associated surface
+      contactsFound_.insert(contact.second.id());
+    }
+  }
+}
+
+template<typename ContactT>
+void ContactsManager<ContactT>::findContactsFromSensors(const mc_control::MCController & ctl,
+                                                        const std::string & robotName)
+{
+  findContactsFromSurfaces(ctl, robotName);
+}
+
+template<typename ContactT>
+std::string ContactsManager<ContactT>::set_to_string(const ContactsSet & contactSet)
+{
+  if(contactSet.cbegin() == contactSet.cend()) { return ""; }
+  std::ostringstream out;
+  out.precision(std::numeric_limits<int>::digits10);
+  out << std::fixed << getNameFromIdx(*contactSet.cbegin());
+
+  for(auto it = std::next(contactSet.cbegin()); it != contactSet.cend(); ++it)
+  {
+    out << ", ";
+    out << std::fixed << getNameFromIdx(*it);
+  }
+  return out.str();
+}
+
+} // namespace mc_state_observation::measurements
```

Comment:
I have pushed something closer to what I had in mind about passing callbacks to the update function: https://github.com/gergondet/mc_state_observation/commit/ad5f44762e35e13e5fa530726855ae2b12ddb0fb

(matching observer code change: https://github.com/gergondet/mc_state_observation/commit/1dbe83030ec86c991888825f27379e81fbfd7894 without considering whether going through the contact list is absolutely needed or it can be done in the callback)